src/content/en/fundamentals/engage-and-retain/push-notifications/sending-messages.markdown

```diff
@@ -3,93 +3,343 @@ layout: shared/narrow
+subscribing browser followed by unique identifier for the user. This is called a
+subscription ID or  a registration ID. This tells your server how to identify
+you the messaging  server.
+  
+**keys**—Encryption keys used for encrypting data passed to the service worker 
+messages. It contains the following:
+
+* **auth**-A 16 byte authentication secret generated by the browser.
+* **p256dh**-A 65 byte containing the vapidPublicKey created on your app server.
+
+**Note:** In many of the relevant specs bytes are called octets. The term is 
+used because of legacy and communication systems where bytes may not always be 8 
+bits long.
+
+## Creating the message
+
+This is where things start to get a little crazy. In this section we're no 
+longer in your client app. We're on the app server where we are going to create 
+and send a message to the client. There's much to keep track of. 
+
+Before continuing, let's review what we have and where it came from.
+
+* **Subscription object** - his came from the client. It contains the endpoint 
+  of the messaging server, a copy of the public key (`applicationServerKey`), and an authentication 
+  secret generated by the client. From here on, we're going to stop talking 
+  about the subscription object and only refer to the **endpoint**, **public 
+  key**, and **authentication secret**.
+* **private key** - The private key corresponding to the public key. This is the private key you created in 'Generating the applicationServerKey'.
+
+We're going to look at message creation in three parts. First we'll create some 
+HTTP headers, then we'll create a payload for the message, and finally we'll 
+combine them and send them to the messaging server.
+
+### A note about code samples
+
+The code samples given in this section, which are taken from the [web-push node 
+library](https://github.com/web-push-libs/web-push) are not ready to use as code 
+samples frequently and ideally should be. Given the complexity of this subject, 
+we felt it was better to describe it at a conceptual level and restrict sample 
+codes to snipits that illustrate those concepts.
+
+### The product
+
+Let's look at what we're going to end up with, then we'll talk about how to 
+build it. 
+
+<pre>POST /push-service/send/dbDqU8xX10w:APA91b... HTTP/1.1  
+Host: push.example.net  
+Push-Receipt: https://push.example.net/r/3ZtI4YVNBnUUZhuoChl6omU  
+TTL: 43200  
+Content-Type: text/plain;charset=utf8  
+Content-Length: 36  
+Authorization: Bearer 
+eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJhdWQiOiJodHRwczovL3B1c2guZXhhbXBsZS5uZXQiLCJleHAiOjE0NTM1MjM3NjgsInN1YiI6Im1haWx0bzpwdXNoQGV4YW1wbGUuY29tIn0.i3CYb7t4xfxCDquptFOepC9GAu\_HLGkMlMuCGSK2rpiUfnK9ojFwDXb1JrErtmysazNjjvW2L9OkSSHzvoD1oA  
+Crypto-Key: 
+p256ecdsa=BA1Hxzyi1RUM1b5wjxsn7nGxAszw2u61m164i3MrAIxHF6YK5h4SDYic-dRuU\_RCPCfA5aq9ojSwk5Y2EmClBPsiChYuI3jMzt3ir20P8r\_jgRR-dSuN182x7iB</pre>
+
+It should be fairly obvious how the endpoint of the subscription object is used. 
```

Comment:
Just say "First thing to call out is we are sending a request to the endpoint from the subscription object".

Developers who have done GCM stuff will not be doing this and for new comers, it's not obvious what you are calling out.